### PR TITLE
[WB-4604] Artifact checkout API

### PR DESCRIPTION
**Pull request by annirudh**
https://wandb.atlassian.net/browse/WB-4604

Description
-----------

Adds a new `checkout()` artifact API that mimics `git checkout` by replacing the contents of a specified directory with the contents of the artifact.

Testing
-------

- Ran loadtest with the checkout option (verify now errors if there are mismatched files)

### CODE REVIEW

**Review by vanpelt — CHANGES_REQUESTED** [review]
This should be pretty easy to test.  See, `wandb_artifacts_test.py`

**Comment by annirudh** [comment]
> This should be pretty easy to test. See, `wandb_artifacts_test.py`

@vanpelt Added!

**Comment by vanpelt** [comment]
Nice, deciphering the [code cov report](https://app.codecov.io/gh/wandb/client/compare/1901/diff) can be a bit of a mystery.  We aggressively try to ensure anything in sdk stays covered.  I *think adding a test for [this case](https://app.codecov.io/gh/wandb/client/compare/1901/diff#D3L453) will fix the current codecov check.  @raubitsj seems like testing [this case](https://app.codecov.io/gh/wandb/client/compare/1901/diff#D4L589) isn't something we should need...

File: wandb/apis/public.py [hunk 1]
```diff
@@ -2914,6 +2914,55 @@ def download(self, root=None, recursive=False):
...
+
+        for root, _, files in os.walk(dirpath):
+            for file in files:
+                full_path = os.path.join(root, file)
+                artifact_path = util.to_forward_slash_path(
+                    os.path.relpath(full_path, start=dirpath)
+                )
+                try:
+                    self.get_path(artifact_path)
+                except KeyError:
```
Comment (tssweeney) [inline wandb/apis/public.py:2947]: Nit: this block is shared with the checkout method above - could be refactored to accept a callback when there is a KeyError.

**Review by tssweeney — APPROVED** [review]
LGTM - awesome add

**Review by vanpelt — APPROVED** [review]

**Pull request merged**